torch_np/_detail/_util.py

```diff
@@ -0,0 +1,292 @@
+"""Assorted utilities, which do not need anything other then torch and stdlib.
+"""
+
+import operator
+
+import torch
+
+from . import _scalar_types
+
+# https://github.com/numpy/numpy/blob/v1.23.0/numpy/distutils/misc_util.py#L497-L504
+def is_sequence(seq):
+    if isinstance(seq, str):
+        return False
+    try:
+        len(seq)
+    except Exception:
+        return False
+    return True
+
+
+def subok_not_ok(like=None, subok=False):
+    if like is not None:
+        raise ValueError("like=... parameter is not supported.")
+    if subok:
+        raise ValueError("subok parameter is not supported.")
+
+
+class AxisError(ValueError, IndexError):
+    pass
+
+
+class UFuncTypeError(TypeError, RuntimeError):
+    pass
+
+
+# a replica of the version in ./numpy/numpy/core/src/multiarray/common.h
+def normalize_axis_index(ax, ndim, argname=None):
+    if ax < -ndim or ax >= ndim:
+        raise AxisError(f"axis {ax} is out of bounds for array of dimension {ndim}")
+    if ax < 0:
+        ax += ndim
+    return ax
+
+
+# from https://github.com/numpy/numpy/blob/main/numpy/core/numeric.py#L1378
+def normalize_axis_tuple(axis, ndim, argname=None, allow_duplicate=False):
+    """
+    Normalizes an axis argument into a tuple of non-negative integer axes.
+    This handles shorthands such as ``1`` and converts them to ``(1,)``,
+    as well as performing the handling of negative indices covered by
+    `normalize_axis_index`.
+    By default, this forbids axes from being specified multiple times.
+    Used internally by multi-axis-checking logic.
+    .. versionadded:: 1.13.0
```

Comment:
minor: this docstring misses a bunch of blank lines

Reply:
Restored.